### DISCUSSION TITLE
Integration with Claris FileMaker Custom App Upgrade Tool

### DISCUSSION THREAD

**Discussion by marcel-more** [fmIDE/fmIDE#21] [Ideas]
fmeIDE is the ideal starting point for any structured developer interaction with a FileMaker solutions code base.
One limiting fact right now are the restrictions by FileMaker Pros interface itself that does not allow scripted access to all kind of objects and structures behind the scenes.

Fortunately Claris offers an approach to apply XML based patches to a given solution.
The 'FileMaker Custom App Upgrade Tool' allows to deploy XML snippets that are merged with a FileMaker file to provide a structured upgrade mechanism:
https://help.claris.com/en/app-upgrade-tool-guide/

Perhaps it makes sense to expand fmIDEs concept in a way that modifications are not only applied directly into an opened solution file but instead it prepares an XML patch file that is then automatically applied to the solution on the command line with help of the 'FileMaker Custom App Upgrade Tool'. fmIDE could manage to coordinate the whole process of building the patch, closing the files, applying the patch to the selected files, archiving the old version and open the new version of the solution file again. This is initially useful as a tool for local files. Later it could even be extended to hosted files as well.

**Comment by mrwatson-de** [comment]
Exactly!

- the app-upgrade tool has been on my radar a long time and throughout the development of fmIDE (and fmAutoMate), and it is most certainly a tool which will feature in the future within the functionality of fmIDE!

It may well be, that with enough metadata available about the incoming changes and the target database, it would be possible to construct a patch file, that intelligently updates the target file.

Maybe even a GitHub action could automate the process

These are massive dreams!

But then again fmIDE is a massive dream too, ... so step for step forwards!

👍

**Reply by marcel-more** [reply]
Yes, of course... dreaming and getting forward step by step...
You did a great job already to provide a map of possibilities and even deployed a tool that makes sense of it!
This might be a solid foundation for further things to come 🙂👍
